bdk_chain/src/file_store.rs

```diff
@@ -24,52 +24,85 @@ where
-                            if inner.kind() != io::ErrorKind::UnexpectedEof {
-                                return Err(inner);
-                            }
-                        }
-                        true
-                    }
-                };
+    /// Iterates over the stored changeset from first to last changing the seek position at each
+    /// iteration.
+    ///
+    /// The iterator may fail to read an entry and therefore return an error. However the first time
+    /// it returns an error will be the last. After doing so the iterator will always yield `None`.
+    ///
+    /// **WARNING**: This method changes the write position in the underlying file. You should
+    /// always iterate over all entries until `None` is returned if you want your next write to go
+    /// at the end, otherwise you writing over existing enties.
+    pub fn iter_changesets(&mut self) -> Result<EntryIter<'_, KeychainChangeSet<K, P>>, io::Error> {
+        self.db_file.rewind()?;
+
+        Ok(EntryIter::new(&mut self.db_file))
+    }
 
-            if failed {
-                db_file.seek(io::SeekFrom::Start(pos))?;
-                break;
+    /// Loads all the changesets that have been stored as one giant changeset.
+    ///
+    /// This function returns a tuple of the aggregate changeset and a result which indicates
+    /// whether an error occurred while reading or deserializing one of the entries. If so the
+    /// changeset will consist of all of those it was able to read.
+    ///
+    /// You should usually check the error. In many applications it may make sense to do a full
+    /// wallet scan with a stop gap after getting an error since it is likely that one of the
+    /// changesets it was unable to read changed the derivation indicies of the tracker.
+    ///
+    /// **WARNING**: This method changes the write position of the underlying file. The next
+    /// changeset will be written over the erroring entry (or the end of the file if none existed).
+    pub fn aggregate_changeset(&mut self) -> (KeychainChangeSet<K, P>, Result<(), IterError>) {
+        let mut changeset = KeychainChangeSet::default();
+        let result = (|| {
+            let iter_changeset = self.iter_changesets()?;
+            for next_changeset in iter_changeset {
+                changeset.append(next_changeset?);
             }
-        }
+            Ok(())
+        })();
 
-        Ok(Self {
-            deriviation_index_cache: tracker.txout_index.derivation_indices(),
-            db_file,
-            chain_index: Default::default(),
-        })
+        (changeset, result)
     }
 
+    /// Reads and applies all the changesets stored sequentially to tracker, stopping when it fails
+    /// to read the next one.
+    ///
+    /// **WARNING**: This method changes the write position of the underlying file. The next
+    /// changeset will be written over the erroring entry (or the end of the file if none existed).
```

Comment:
The dangers of doing this without truncating invalid `ChangeSet`s may result in loading an invalid tracker.

Imagine the following scenario:

* Changeset 21 is invalid data (but changeset 22 is valid), so we load tracker from changesets 0 to 20.
* The applications appends a changeset (21b) that is exactly the same size as (21).
* Application closes.
* Application reopens and loads tracker with changesets 0 to 22 (inclusive of 21b). It is possible at this point, that loading 21b and 22 in sequence results in inconsistent tracker.

However, we can't just truncate automatically when there is invalid data as that is dangerous. I think the best way to do this is to have a third field in `KeychainStore` that records whether there has been an error and at which file position. We add another method that can truncate problematic portions of the file (that also clears the error field). If the error field is some, many methods will fail.

Reply:
As discussed with @LLFourn, instead of adding a field error (in case of opening up a wrong file), we will introduce some magic bytes to the start of the file, thus reducing the risk of accidental file truncation.